### ISSUE TITLE
Auth fails with "retry function failed after 1 attempt: failed to generate Google Cloud access token"

### ISSUE THREAD

**Issue by PaulRudin** [google-github-actions/auth#243]
### TL;DR

I've tried various config, but I can't get past this error.

### Expected behavior

Successful token creation

### Observed behavior

I get the error: "retry function failed after 1 attempt: failed to generate Google Cloud access token"

### Action YAML

```yaml
name: Tests

on:
  push:
    branches-ignore:
      - main

jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: 'write'
      id-token: 'write'
    steps:
    - uses: actions/checkout@v3
    - name: Set up Python
      uses: actions/setup-python@v2
      with:
        python-version: 3.10.2
    - name: Install poetry
      uses: abatilo/actions-poetry@v2.0.0
      with:
        poetry-version: 1.2.2

    - id: 'auth'
      uses: 'google-github-actions/auth@v1'
      with:
        token_format: 'access_token'
        workload_identity_provider: 'projects/xxx/locations/global/workloadIdentityPools/github-pool/providers/github-provider'
        service_account: 'xxx@yyy.iam.gserviceaccount.com'

    - name: install dependencies
      run: poetry install

    - name: Run tests
      run: poetry run pytest --mypy --cov-report=term --cov-report=xml:cov.xml --cov
      
    - name: Code Coverage Report
      uses: irongut/CodeCoverageSummary@v1.3.0
      with:
        filename: cov.xml

    - name: Pytest coverage comment
      id: coverageComment
      uses: MishaKav/pytest-coverage-comment@main
      with:
        pytest-xml-coverage-path: cov.xml

    - name: Update Readme with Coverage Html
      run: |
        sed -i '/<!-- Pytest Coverage Comment:Begin -->/,/<!-- Pytest Coverage Comment:End -->/c\<!-- Pytest Coverage Comment:Begin -->\${{ steps.coverageComment.outputs.coverageHtml }}${{ steps.coverageComment.outputs.summaryReport }}<!-- Pytest Coverage Comment:End -->' ./README.md
    - name: Commit & Push changes to Readme
      uses: stefanzweifel/git-auto-commit-action@v4
      with:
        file_pattern: README.md
```

### Log output

```text
Run google-github-actions/auth@v1
  with:
    token_format: access_token
    workload_identity_provider: projects/xxx/locations/global/workloadIdentityPools/github-pool/providers/github-provider
    service_account: xxx@yyy.iam.gserviceaccount.com
    create_credentials_file: true
    export_environment_variables: true
    cleanup_credentials: true
    access_token_lifetime: 3600s
    access_token_scopes: https://www.googleapis.com/auth/cloud-platform
    retries: 0
    id_token_include_email: false
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.10.2/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.2/x64/lib
Created credentials file at "/home/runner/work/xxx/xxx/gha-creds-4f9276f1969ce354.json"
Error: google-github-actions/auth failed with: retry function failed after 1 attempt: failed to generate Google Cloud access token for xxx@yyy.iam.gserviceaccount.com: (403) {
  "error": {
    "code": 403,
    "message": "The caller does not have permission",
    "status": "PERMISSION_DENIED"
  }
}
```

### Additional information

I've used the gh-oidc terraform module to set things up; and that applies without errors, but presumably there's some misconfiguration somewhere - but I'm not sure what I should be looking for.

https://registry.terraform.io/modules/terraform-google-modules/github-actions-runners/google/latest/submodules/gh-oidc

**Comment by sethvargo** [comment]
Hi @PaulRudin - please review the [troubleshooting steps](https://github.com/google-github-actions/auth/blob/main/docs/TROUBLESHOOTING.md). There are many common misconfigurations that cause that error.

**Comment by PaulRudin** [comment]
Yup, I read that. I guess the fact that there are many potential misconfigurations that makes it tricky to spot the problem.

**Comment by sethvargo** [comment]
Did you provide sa_mappings in that Terraform configuration?

**Comment by PaulRudin** [comment]
Yeah, that block is defined by:
```jsonnet
    sa_mapping: {
	"github-fplscore-sa" : {
	  sa_name: "${google_service_account.github_fplscore_sa.name}",
	  attribute: "attribute.repository/PaulRudin/fplscore",
	},
	"github-fplserve-sa": {
	  sa_name: "${google_service_account.github_fplserve_sa.name}",
	  attribute: "attribute.repository/PaulRudin/fplserve",
	}
      }
```

**Comment by PaulRudin** [comment]
I tried enabling audit logging as suggested in the troubleshooting guide, but I couldn't see any log entries that seem relevant after rerunning the workflow. But I'm not sure what I should be looking for.

**Comment by djfinnoy** [comment]
I'm experiencing the same problem. Service account impersonation is not working, been trying for two days now.
I've tried all the steps in the troubleshooting docs, and in closed issues; I simply cannot limit a service account to any attributes, or even the `subject` field.
I'm able to get around this issue by using conditions on the provider; and allowing service accounts to access "all identities in the pool".  This effectively limits me to one service account per provider, which I guess is fine for my use-case.

I'm pretty confident that something is wrong with service account impersonations. I can't find anything useful in the logs tho. To reproduce, try to limit a service account to `repository_owner` with the latest version of the auth workflow. I just set up the pool using my browser as a test; everything seems to be in order, and the `principalSet` IAM rules generated by console actions seem correct, but the action still fails with `failed to generate Google Cloud access token`.

**Comment by PaulRudin** [comment]
So, based on that I just used a single service account and used "*" as the attribute in the sa_mapping, and now the authentication seems to succeed.

**Comment by sethvargo** [comment]
Hi @PaulRudin @djfinnoy 

Did you provide any attribute mappings in the configuration (Terraform)? See step 3 of the troubleshooting guide, but you must have an attribute mapping for any attribute conditions. You cannot create an attribute condition unless you map values from the incoming GitHub OIDC token. In that Terraform module, it would be something like:

```terraform
module "gh_oidc" {
  source      = "terraform-google-modules/github-actions-runners/google//modules/gh-oidc"
  # ...

  attribute_mapping = {
    "google.subject": "assertion.sub"
    "attribute.repository" : "assertion.repository"
  }

  sa_mapping = {
    "foo-service-account" = {
      sa_name   = "projects/my-project/serviceAccounts/foo-service-account@my-project.iam.gserviceaccount.com"
      attribute = "attribute.repository/user/repo"
    }
  }
}
```

Also double check capitalization on the repository owner and repository names.

**Comment by lucasmellos** [comment]
Facing the same error on my side. After upgrading to v1 it broke my pipeline.

**Comment by sethvargo** [comment]
Hi @lucasmellos - there were [only documentation changes between v0 and v1](https://github.com/google-github-actions/auth/compare/v0.8.3...v1.0.0), so I'm not sure how that would affect your pipeline.

**Comment by lucasmellos** [comment]
On my side, it was the lack of SA attachment. Notting related to this action.

**Comment by sethvargo** [comment]
Hi @djfinnoy - it's also possible that you're not waiting ~5min in between changes. WIF pools take a long time to provision and are eventually consistent, especially configuration changes. Attribute conditions should be `attribute.something`.

**Comment by PaulRudin** [comment]
> Hi @PaulRudin @djfinnoy
> 
> Did you provide any attribute mappings in the configuration (Terraform)? See step 3 of the troubleshooting guide, but you must have an attribute mapping for any attribute conditions. You cannot create an attribute condition unless you map values from the incoming GitHub OIDC token. In that Terraform module, it would be something like:
> 
> ```terraform
> module "gh_oidc" {
>   source      = "terraform-google-modules/github-actions-runners/google//modules/gh-oidc"
>   # ...
> 
>   attribute_mapping = {
>     "google.subject": "assertion.sub"
>     "attribute.repository" : "assertion.repository"
>   }
> 
>   sa_mapping = {
>     "foo-service-account" = {
>       sa_name   = "projects/my-project/serviceAccounts/foo-service-account@my-project.iam.gserviceaccount.com"
>       attribute = "attribute.repository/user/repo"
>     }
>   }
> }
> ```
> 
> Also double check capitalization on the repository owner and repository names.

I didn't specify the attribute mapping explicitly, but that terraform module has a default that includes the mapping you suggest: https://github.com/terraform-google-modules/terraform-google-github-actions-runners/blob/760e9710da0a2378d2acbb58fb4c06c007b4ec79/modules/gh-oidc/variables.tf#L68

**Comment by sethvargo** [comment]
@PaulRudin understood. Unfortunately without more information, it's difficult to debug this. Have you enabled the Security Token Service Activity logs for Admin, Data Read, and Data Write and looked at the logs?

**Comment by sethvargo** [comment]
Hi @PaulRudin - can you share some screenshots of your Pool and Provider configuration please?

**Comment by PaulRudin** [comment]
As I mentioned I've now switch to a single service account and using '*' for the attribute and things work on that basis. I'll try to make a minimal reproducible example without messing with my existing setup.

**Comment by sethvargo** [comment]
Hey @PaulRudin - any luck on building a reproduction case?
